src/omv/img/draw.c

```diff
@@ -474,55 +488,152 @@ static int safe_map_pixel(int dst_bpp, int src_bpp, int pixel)
-                    : safe_map_pixel(img_bpp, other_bpp, other_pixel);
-                int img_pixel = imlib_get_pixel(img, x_off + x, y_off + y);
-				
-                int result_pixel;
-                switch (img_bpp) {
-                    case IMAGE_BPP_BINARY: {
-                        result_pixel = (other_pixel*alpha + img_pixel*neg_alpha)>=0.5?1:0;
-                        break;
-                    }
-                    case IMAGE_BPP_GRAYSCALE: {
-                        result_pixel = other_pixel*alpha + img_pixel*neg_alpha;
-                        break;
+
+    // Left or top of other is out of bounds
+    int other_x_start = (x_off < 0) ? -x_off : 0;
+    int other_y_start = (y_off < 0) ? -y_off : 0;
+    
+    // Scaled other size
+    const int other_width_scaled = fast_floorf(abs(other->w * x_scale));
+    const int other_height_scaled = fast_floorf(abs(other->h * y_scale));
+
+    // Right or bottom of image is out of bounds
+    int other_x_end = (x_off + other_width_scaled >= img->w) ? img->w - x_off : other_width_scaled;
+    int other_y_end = (y_off + other_height_scaled >= img->h) ? img->h - y_off : other_height_scaled;
+
+    // Check bounds are within img
+    if (other_x_start + x_off >= img->w || other_y_start + y_off >= img->h) return;
+    if (other_x_end + x_off <= 0 || other_y_end + y_off <= 0) return;
+
+    // If scaling is negative we essentially flip the other coordinates so they work from bottom right instead of top left
+    if (over_xscale < 0) {
+        other_x_start -= other_width_scaled;
+        other_x_end -= other_width_scaled;
+        x_off += other_width_scaled;
+    }
+
+    if (over_yscale < 0) {
+        other_y_start -= other_height_scaled;
+        other_y_end -= other_height_scaled;
+        y_off += other_height_scaled;
+    }
+
+    switch(img_bpp) {
+        case IMAGE_BPP_BINARY: {
+            // Iterate the img area to be updated
+            for (int y = other_y_start; y < other_y_end; y++) {
+                uint32_t *img_row_ptr = IMAGE_COMPUTE_BINARY_PIXEL_ROW_PTR(img, y_off + y);
+
+                const int other_y = fast_floorf(y * over_yscale);
+                void *other_row_ptr = imlib_compute_row_ptr(other, other_y);
+
+                for (int x = other_x_start; x < other_x_end; x++) {
+                    const int other_x = fast_floorf(x * over_xscale);
+
+                    if ((!mask) || image_get_mask_pixel(mask, other_x, other_y)) {
+                        const uint32_t other_pixel = safe_map_pixel(IMAGE_BPP_BINARY, other_bpp, imlib_get_pixel_fast(other_bpp, other_row_ptr, other_x));
+                        
+                        uint32_t result_pixel;
+                        if (alpha==256) {
+                            result_pixel = other_pixel;
+                        }
+                        else {
+                            const uint32_t img_pixel = IMAGE_GET_BINARY_PIXEL_FAST(img_row_ptr, x_off + x);
```

Comment:
So does this sound correct in the case of alpha < 256?
result_pixel = img_pixel & other_pixel;

Reply:
No, it would be like:

result_pixel = alpha < 128 ? img_pixel :  other_pixel;

I'm not sure if that order is correct given what happens when alpha gets low. Reverse it as required.

Reply:
If alpha<128 we are just copying the image to the image again, so I've just bypassed the draw_image for bitmap if alpha<128

Reply:
Also, the original alpha blending code came from fir_draw_ir, which I'm thinking of changing to utilise draw_image which would save over 300 bytes of flash, if there's a shortage.

Reply:
I have some new alpha blending code which increases performance by 10%, it's a bit hampered by the byte ordering of the RGB565 images.  I'm assuming the ordering is what comes out of the camera.  Would it be a huge task to put it in a sensible order?  I'm guessing you'd just adjust the RGB565 macros, and change the order as it comes out of the camera would be about all there is to do?

Reply:
That shipped sailed years ago. I wanted to fix this back about 5 years ago and it didn't happen. The OV2640/OV5640 have the ability to reverse this. But, for the OV7725 you don't, The byte order comes out of the camera this way.

Also, our code optimizer Larry doesn't use the macros so we'd have to go back and fix a lot of his code he wrote.

It could be done... but, it would need to be done all at once. Every code path. So, not going to happen without a lot more folks working on the code.